Exploring ebay Car Sales Data

In [1]:
import pandas as pd
import numpy as np
import csv

In [2]:
autos = pd.read_csv('autos.csv',encoding='Latin-1')

In [3]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [4]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

There are 5 integer(64) types with the remaining 15 being objects.

In [5]:


copy=autos.columns
print(copy)#display initial column names
new_cols=[]
for col in copy:# iterating through the column_names
    if col=='yearOfRegistration':
        new_cols.append('registration_year')
    elif col=='monthOfRegistration':
        new_cols.append('registration_month')
    elif col=='notRepairedDamage':
        new_cols.append('unrepaired_damage')
    elif col=='dateCreated':
        new_cols.append('ad_created')
    else:
        new_cols.append(col.lower())
                       
autos.columns=new_cols
autos.head()                       
                       
                       


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


datecrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicletype  registration_year  \
0  privat   Angebot  $5,000  control         bus               2004   
1  privat   Angebot  $8,500  control   limousine               1997   
2  privat   Angebot  $8,990     test   limousine               2009   
3  privat   Angebot  $4,350  control  kleinwagen               2007   
4  privat   Angebot  $1,350     test       kombi               2003   

     gearbox  powerps   model   odometer  registration_month fueltype  \
0    manuell      158  andere  150,000km                   3      lpg   
1  automatik      286     7er  150,000km                   6   benzin   
2    manuell      102    golf   70,000km                   7   benzin   
3  automatik       71  fortwo   70,000km                   6   benzin   
4    manuell        0   focus  150,000km                   7   benzin   

        brand unrepaired_damage           ad_created  nrofpictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalcode             lastseen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Changed the column names for clarity.

In [6]:
autos.describe(include='all')

datecrawled         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicletype  registration_year  gearbox       powerps  model  \
count        44905       50000.000000    47320  50000.000000  47242   
unique           8                NaN        2           NaN    245   
top      limousine                NaN  manuell           NaN   golf   
freq         12859                NaN    36993           NaN   4024   
mean           NaN        2005.073280      NaN    116.355920    NaN   
std            NaN         105.712813      NaN    209.216627    NaN   
min            NaN        1000.000000      NaN      0.000000    NaN   
25%            NaN        1999.000000      NaN     70.000000    NaN   
50%            NaN        2003.000000      NaN    105.000000    NaN   
75%            NaN        2008.000000      NaN    150.000000    NaN   
max            NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fueltype       brand unrepaired_damage  \
count       50000        50000.000000    45518       50000             40171   
unique         13                 NaN        7          40                 2   
top     150,000km                 NaN   benzin  volkswagen              nein   
freq        32424                 NaN    30107       10687             35232   
mean          NaN            5.723360      NaN         NaN               NaN   
std           NaN            3.711984      NaN         NaN               NaN   
min           NaN            0.000000      NaN         NaN               NaN   
25%           NaN            3.000000      NaN         NaN               NaN   
50%           NaN            6.000000      NaN         NaN               NaN   
75%           NaN            9.000000      NaN         NaN               NaN   
max           NaN           12.000000      NaN         NaN               NaN   

                 ad_created  nrofpictures    postalcode             lastseen  
count                 50000       50000.0  50000.000000                50000  
unique                   76           NaN           NaN                39481  
top     2016-04-03 00:00:00           NaN           NaN  2016-04-07 06:17:27  
freq                   1946           NaN           NaN                    8  
mean                    NaN           0.0  50813.627300                  NaN  
std                     NaN           0.0  25779.747957                  NaN  
min                     NaN           0.0   1067.000000                  NaN  
25%                     NaN           0.0  30451.000000                  NaN  
50%                     NaN           0.0  49577.000000                  NaN  
75%                     NaN           0.0  71540.000000                  NaN  
max                     NaN           0.0  99998.000000                  NaN

Numeric data stored as text;odometer, price.
Columns that require further investigation; nrofpictures.

In [7]:
autos['price']=autos['price'].str.replace(',','')
autos['price']=autos['price'].str.replace('$','')
autos['price']=autos['price'].astype(int)

autos['odometer']=autos['odometer'].str.replace(',','')
autos['odometer']=autos['odometer'].str.replace('km','')
autos['odometer']=autos['odometer'].astype(int)

autos.rename({'odometer':'odometer_km'},axis=1,inplace=True)

In [8]:
autos.head(1)

datecrawled                              name  seller offertype  \
0  2016-03-26 17:47:46  Peugeot_807_160_NAVTECH_ON_BOARD  privat   Angebot   

   price   abtest vehicletype  registration_year  gearbox  powerps   model  \
0   5000  control         bus               2004  manuell      158  andere   

   odometer_km  registration_month fueltype    brand unrepaired_damage  \
0       150000                   3      lpg  peugeot              nein   

            ad_created  nrofpictures  postalcode             lastseen  
0  2016-03-26 00:00:00             0       79588  2016-04-06 06:45:54

In [12]:
autos['odometer_km'].unique().shape[0]

13

In [11]:
autos['price'].unique().shape[0]

2357

In [14]:
autos['odometer_km'].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [16]:
autos['odometer_km'].value_counts().sort_values()

10000       264
20000       784
30000       789
40000       819
5000        967
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [17]:
autos['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [21]:
autos['price'].value_counts().sort_values()

33980       1
19599       1
7333        1
17899       1
66          1
         ... 
1000      639
2500      643
1500      734
500       781
0        1421
Name: price, Length: 2357, dtype: int64

In [24]:
autos['datecrawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.02538
2016-03-06    0.01394
2016-03-07    0.03596
2016-03-08    0.03330
2016-03-09    0.03322
2016-03-10    0.03212
2016-03-11    0.03248
2016-03-12    0.03678
2016-03-13    0.01556
2016-03-14    0.03662
2016-03-15    0.03398
2016-03-16    0.02950
2016-03-17    0.03152
2016-03-18    0.01306
2016-03-19    0.03490
2016-03-20    0.03782
2016-03-21    0.03752
2016-03-22    0.03294
2016-03-23    0.03238
2016-03-24    0.02910
2016-03-25    0.03174
2016-03-26    0.03248
2016-03-27    0.03104
2016-03-28    0.03484
2016-03-29    0.03418
2016-03-30    0.03362
2016-03-31    0.03192
2016-04-01    0.03380
2016-04-02    0.03540
2016-04-03    0.03868
2016-04-04    0.03652
2016-04-05    0.01310
2016-04-06    0.00318
2016-04-07    0.00142
Name: datecrawled, dtype: float64

In [26]:
autos['ad_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2015-06-11    0.00002
2015-08-10    0.00002
2015-09-09    0.00002
2015-11-10    0.00002
2015-12-05    0.00002
               ...   
2016-04-03    0.03892
2016-04-04    0.03688
2016-04-05    0.01184
2016-04-06    0.00326
2016-04-07    0.00128
Name: ad_created, Length: 76, dtype: float64

In [28]:
autos['lastseen'].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.00108
2016-03-06    0.00442
2016-03-07    0.00536
2016-03-08    0.00760
2016-03-09    0.00986
2016-03-10    0.01076
2016-03-11    0.01252
2016-03-12    0.02382
2016-03-13    0.00898
2016-03-14    0.01280
2016-03-15    0.01588
2016-03-16    0.01644
2016-03-17    0.02792
2016-03-18    0.00742
2016-03-19    0.01574
2016-03-20    0.02070
2016-03-21    0.02074
2016-03-22    0.02158
2016-03-23    0.01858
2016-03-24    0.01956
2016-03-25    0.01920
2016-03-26    0.01696
2016-03-27    0.01602
2016-03-28    0.02086
2016-03-29    0.02234
2016-03-30    0.02484
2016-03-31    0.02384
2016-04-01    0.02310
2016-04-02    0.02490
2016-04-03    0.02536
2016-04-04    0.02462
2016-04-05    0.12428
2016-04-06    0.22100
2016-04-07    0.13092
Name: lastseen, dtype: float64

In [29]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Registration year is stored as float 64.
it contains absurd entries. like 9999 and 1000 as entries. cars weren't invented as of the year 1000 and 9999 is way into the future.

In [30]:
autos['registration_year'].value_counts()

2000    3354
2005    3015
1999    3000
2004    2737
2003    2727
        ... 
1931       1
1929       1
1001       1
9996       1
1952       1
Name: registration_year, Length: 97, dtype: int64

the minimum acceptable registration year should be 1800 seeing as that was when cars were invented. the maximum will be 2023; the most recent year. 

In [39]:
autos.loc[(autos['registration_year']>1800)&(autos['registration_year']<2023),'registration_year'].value_counts(normalize=True)

2000    0.067112
2005    0.060329
1999    0.060029
2004    0.054766
2003    0.054566
          ...   
1943    0.000020
1929    0.000020
1938    0.000020
1939    0.000020
1952    0.000020
Name: registration_year, Length: 81, dtype: float64

As shown above, vehicles registered in the year 2000 are more.

Exploring Price by Brand

i have chosen to choose the top 20 brands.

In [41]:
top_20= autos['brand'].value_counts(normalize=True).head(20)

In [57]:
top_20.index[1]

'opel'

In [69]:
brand_price_agg={}
brand_mileage={}
for brand in top_20.index:
    price_mean = autos.loc[autos['brand']==brand,'price'].mean()
    mileage_mean=autos.loc[autos['brand']==brand,'odometer_km'].mean()
    brand_price_agg[brand]=round(price_mean,2)
    brand_mileage[brand]=round(mileage_mean,2)

In [62]:
print(brand_price_agg)

{'volkswagen': 6384.17, 'opel': 5106.09, 'bmw': 8252.92, 'mercedes_benz': 29511.96, 'audi': 8965.56, 'ford': 7105.66, 'renault': 2351.3, 'peugeot': 3010.87, 'fiat': 12134.21, 'seat': 4219.43, 'skoda': 6305.04, 'mazda': 3962.54, 'nissan': 4588.88, 'smart': 3482.97, 'citroen': 42657.46, 'toyota': 5097.94, 'sonstige_autos': 38300.84, 'hyundai': 5316.75, 'volvo': 31689.91, 'mini': 10392.39}


In [66]:
print('{} has the highest price average of {}'.format(max(brand_price_agg),brand_price_agg[max(brand_price_agg)]))

volvo has the highest price average of 31689.91


In [68]:
print('{} has the lowest price average of {}'.format(min(brand_price_agg),brand_price_agg[min(brand_price_agg)]))

audi has the lowest price average of 8965.56


In [70]:
print(brand_mileage)

{'volkswagen': 128955.27, 'opel': 129298.66, 'bmw': 132521.64, 'mercedes_benz': 130886.14, 'audi': 129643.94, 'ford': 124131.93, 'renault': 128223.79, 'peugeot': 127352.34, 'fiat': 117037.46, 'seat': 122061.64, 'skoda': 110947.84, 'mazda': 125132.1, 'nissan': 118978.78, 'smart': 100756.06, 'citroen': 119764.62, 'toyota': 115988.65, 'sonstige_autos': 87188.64, 'hyundai': 106782.79, 'volvo': 138632.39, 'mini': 89375.0}


In [74]:
mile = pd.Series(brand_mileage)#Convert milege to series

In [75]:
print(mile)

volkswagen        128955.27
opel              129298.66
bmw               132521.64
mercedes_benz     130886.14
audi              129643.94
ford              124131.93
renault           128223.79
peugeot           127352.34
fiat              117037.46
seat              122061.64
skoda             110947.84
mazda             125132.10
nissan            118978.78
smart             100756.06
citroen           119764.62
toyota            115988.65
sonstige_autos     87188.64
hyundai           106782.79
volvo             138632.39
mini               89375.00
dtype: float64


In [76]:
price_avg=pd.Series(brand_price_agg)

In [77]:
print(price_avg)

volkswagen         6384.17
opel               5106.09
bmw                8252.92
mercedes_benz     29511.96
audi               8965.56
ford               7105.66
renault            2351.30
peugeot            3010.87
fiat              12134.21
seat               4219.43
skoda              6305.04
mazda              3962.54
nissan             4588.88
smart              3482.97
citroen           42657.46
toyota             5097.94
sonstige_autos    38300.84
hyundai            5316.75
volvo             31689.91
mini              10392.39
dtype: float64


In [80]:
mileage_price_avg=pd.DataFrame(price_avg,columns=['mean_price'])

In [83]:
mileage_price_avg['mean_mileage']=mile

In [84]:
mileage_price_avg

mean_price  mean_mileage
volkswagen         6384.17     128955.27
opel               5106.09     129298.66
bmw                8252.92     132521.64
mercedes_benz     29511.96     130886.14
audi               8965.56     129643.94
ford               7105.66     124131.93
renault            2351.30     128223.79
peugeot            3010.87     127352.34
fiat              12134.21     117037.46
seat               4219.43     122061.64
skoda              6305.04     110947.84
mazda              3962.54     125132.10
nissan             4588.88     118978.78
smart              3482.97     100756.06
citroen           42657.46     119764.62
toyota             5097.94     115988.65
sonstige_autos    38300.84      87188.64
hyundai            5316.75     106782.79
volvo             31689.91     138632.39
mini              10392.39      89375.00

Finding the most common brand/model combinations

In [88]:
brands = autos['brand'].unique()#get a series of unique brands